README
This package is http://pear.php.net/package/PEAR_PackageFileManager_Frontend and has been migrated from http://svn.php.net/repository/pear/packages/PEAR_PackageFileManager_Frontend

Please report all new issues via the PEAR bug tracker.

If this package is marked as unmaintained and you have fixes, please submit your pull requests and start discussion on the pear-qa mailing list.

To test, run either
$ phpunit tests/
  or
$ pear run-tests -r

To build, simply
$ pear package

To install from scratch
$ pear install package.xml

To upgrade
$ pear upgrade -f package.xml


NEWS
Changes since version 0.7.0 (2006-08-19)

Changes to Frontend.php
  - add support for tasks file : convert line endings (windows and unix)
  - add support for install as name condition
  - remove support of package xml 1.0 generation

Changes since version 0.6.0 (2006-07-24)

Changes to Frontend.php
  - setFileList() method $plugin parameter is no more optional.

Changes since version 0.5.0 (2006-05-29)

Changes to Frontend.php
  - introduces new types (group-package, group-subpackage, group-extension)
    to manage group dependencies
  - be sure to have a default files list in session
  - removed page 6 (roles)  management
  - global var '_PEAR_PACKAGEFILEMANAGER_FRONTEND_SINGLETON' was removed
    to avoid to pollute the global namespace.
    Singleton pattern used a static variable (as it should be since beginning)
  - better support of package xml <group> tag especially with package, subpackage, and extension

Changes since version 0.4.0 (2006-05-21)

Changes to Frontend.php
  * bugs:
  - Avoid to propagate wrong data content to the other package dependencies.
  - php extension from package dependencies are now detected correctly.
  - be sure to returns good data when getDefaults('dependencies') is called.
  - keep good version of pear installer when import a package 2.0
  - just for a new package from scratch pear installer is set to 1.4.3 minimum
    due to vunerability found on Nov 4 2005, rather than default 1.4.0b1

Changes since version 0.3.0 (2006-05-13)

Changes to Frontend.php
  - new api : getPreferences() to retrieve easily user options loaded.
  - options filelistgenerator was renamed to plugingenerator, and new filelistgenerator
    begin equal to PFM same option.
  - new configuration structure (more simple).
  - setOption() and getOption() did not used any more the xpath ability of PEAR::Config
  - realpathnix() and array_merge_recursive() are static class methods
    so they don't pollute the global namespace.
  - loadPreferences() is no more mandatory, and you can reload default options.
  - add a hack for PEAR::Config that cannot parse correctly boolean for XML container.
  - reduces amount of data stored in session

Changes since version 0.2.0 (2006-04-02)

Changes to Frontend.php
  - Add full support for 'group' and 'installconditions' package.xml version 2.0 tags



Frontend/Decorator.php
<?php
/**
 * Decorator for PEAR_PackageFileManager_Frontend
 *
 * PHP versions 4 and 5
 *
 * LICENSE: This source file is subject to version 3.01 of the PHP license
 * that is available through the world-wide-web at the following URI:
 * http://www.php.net/license/3_01.txt.  If you did not receive a copy of
 * the PHP License and are unable to obtain it through the web, please
 * send a note to license@php.net so we can mail you a copy immediately.
 *
 * @category   PEAR
 * @package    PEAR_PackageFileManager_Frontend
 * @author     Laurent Laville <pear@laurent-laville.org>
 * @copyright  2005-2006 Laurent Laville
 * @license    http://www.php.net/license/3_01.txt  PHP License 3.01
 * @version    CVS: $Id$
 * @since      File available since Release 0.1.0
 */

require_once 'PEAR/PackageFileManager/Frontend.php';

/**
 * Decorates any PEAR_PackageFileManager_Frontend class
 *
 * @category   PEAR
 * @package    PEAR_PackageFileManager_Frontend
 * @author     Laurent Laville <pear@laurent-laville.org>
 * @copyright  2005-2006 Laurent Laville
 * @license    http://www.php.net/license/3_01.txt  PHP License 3.01
 * @version    Release: @package_version@
 * @since      Class available since Release 0.1.0
 * @abstract
 */

class PEAR_PackageFileManager_Frontend_Decorator extends PEAR_PackageFileManager_Frontend
{
    /**
     * An instance PEAR_PackageFileManager_Frontend
     *
     * @var    object
     * @since  0.1.0
     */
    var $fe;

    /**
     * Decorator constructor
     *
     * @param  object   $frontend  instance of PEAR_PackageFileManager_Frontend
     * @since  0.1.0
     * @access public
     */
    function PEAR_PackageFileManager_Frontend_Decorator(&$frontend)
    {
        $this->fe =& $frontend;
    }

    /**
     * Decorator::getMaintList()
     *
     * @return array
     * @since  0.1.0
     * @access public
     * @see    PEAR_PackageFileManager_Frontend::getMaintList()
     */
    function getMaintList($users)
    {
        return $this->fe->getMaintList($users);
    }

    /**
     * Decorator::getFileList()
     *
     * @param  boolean  $default  if we get initial data set at first run
     * @param  boolean  $ignore   Either if you want all files or just ignored
     * @param  string   $plugin   PEAR_PackageFileManager filelist generator
     * @return array
     * @since  0.1.0
     * @access public
     * @see    PEAR_PackageFileManager_Frontend::getFileList()
     */
    function getFileList($default, $ignore, $plugin)
    {
        return $this->fe->getFileList($default, $ignore, $plugin);
    }

    /**
     * Decorator::setFileList()
     *
     * @param  string   $plugin   PEAR_PackageFileManager filelist generator
     * @return void
     * @since  0.1.0
     * @access public
     * @see    PEAR_PackageFileManager_Frontend::setFileList()
     */
    function setFileList($plugin)
    {
        $this->fe->setFileList($plugin);
    }

    /**
     * Decorator::getDepList()
     *
     * @return array
     * @since  0.1.0
     * @access public
     * @see    PEAR_PackageFileManager_Frontend::getDepList()
     */
    function getDepList()
    {
        return $this->fe->getDepList();
    }

    /**
     * Decorator::getRoleList()
     *
     * @return array
     * @since  0.1.0
     * @access public
     * @see    PEAR_PackageFileManager_Frontend::getRoleList()
     */
    function getRoleList($default)
    {
        return $this->fe->getRoleList($default);
    }

    /**
     * Decorator::getExceptionList()
     *
     * @return array
     * @since  0.1.0
     * @access public
     * @see    PEAR_PackageFileManager_Frontend::getExceptionList()
     */
    function getExceptionList()
    {
        return $this->fe->getExceptionList();
    }
}
?>

examples/THL_Infos.php
<?php
/**
 * An example script that retrieves informations
 * from PEAR::Text_Highlighter package
 * using PEAR_PackageFileManager_Frontend with no front end.
 *
 * @author    Laurent Laville
 * @package   PEAR_PackageFileManager_Frontend
 * @version   $Id$
 * @license   http://www.php.net/license/3_01.txt  PHP License 3.01
 * @copyright 2006 Laurent Laville
 * @ignore
 */

require_once 'PEAR/PackageFileManager/Frontend.php';

/**
 * This class allow to use PEAR_PackageFileManager_Frontend as backend without any frontend.
 * No end-user action needed.
 * @ignore
 */
class PEAR_PackageFileManager_Frontend_Null extends PEAR_PackageFileManager_Frontend
{
    function PEAR_PackageFileManager_Frontend_Null($driver, $packagedirectory, $pathtopackagefile)
    {
        parent::PEAR_PackageFileManager_Frontend($driver, $packagedirectory, $pathtopackagefile);

        // load all default preferences
        $config = false;
        $this->loadPreferences($config);
    }
}

/**
 * A wrapper for PEAR::Var_Dump package or default php var_dump() function.
 * @ignore
 */
function varDump($var)
{
    $available = PEAR_PackageFileManager2::isIncludeable('Var_Dump.php');
    if ($available) {
        include_once 'Var_Dump.php';
        Var_Dump::display($var, false, array('display_mode' => 'HTML4_Table'));
    } else {
        echo '<pre style="background-color:#eee; color:#000; padding:1em;">';
        var_dump($var);
        echo '</pre>';
    }
}

session_start();

// where to find package sources
$pkgDir = 'c:/PEAR/Text/Text_Highlighter-0.6.9';

$pfmfe =& PEAR_PackageFileManager_Frontend::singleton('Null', $pkgDir);
$pfmfe->setOption('baseinstalldir', 'Text');
if ($pfmfe->hasErrors()) {
    $errors = $pfmfe->getErrors();
    varDump($errors);
    die('exit on Error');
}

echo '<h1>Package Summary </h1>';
$def = $pfmfe->getDefaults('package');
varDump($def);

echo '<h1>Latest Release </h1>';
$def = $pfmfe->getDefaults('release');
varDump($def);

echo '<h1>Maintainers </h1>';
$def = $pfmfe->getMaintList();
varDump($def);

echo '<h1>Dependencies </h1>';
$def = $pfmfe->getDepList();
varDump($def);

// cleansweep session data
$pfmfe->container(true);
session_destroy();
?>

examples/PFM_MaintainersList.php
<?php
/**
 * An example script that retrieves maintainers list
 * from PEAR::PEAR_PackageFileManager package
 * using PEAR_PackageFileManager_Frontend with no front end.
 *
 * @author    Laurent Laville
 * @package   PEAR_PackageFileManager_Frontend
 * @version   $Id$
 * @license   http://www.php.net/license/3_01.txt  PHP License 3.01
 * @copyright 2006 Laurent Laville
 * @ignore
 */

require_once 'PEAR/PackageFileManager/Frontend.php';

/**
 * This class allow to use PEAR_PackageFileManager_Frontend as backend without any frontend.
 * No end-user action needed.
 * @ignore
 */
class PEAR_PackageFileManager_Frontend_Null extends PEAR_PackageFileManager_Frontend
{
    function PEAR_PackageFileManager_Frontend_Null($driver, $packagedirectory, $pathtopackagefile)
    {
        parent::PEAR_PackageFileManager_Frontend($driver, $packagedirectory, $pathtopackagefile);

        // load all default preferences
        $config = false;
        $this->loadPreferences($config);
    }
}

/**
 * A wrapper for PEAR::Var_Dump package or default php var_dump() function.
 * @ignore
 */
function varDump($var)
{
    $available = PEAR_PackageFileManager2::isIncludeable('Var_Dump.php');
    if ($available) {
        include_once 'Var_Dump.php';
        Var_Dump::display($var, false, array('display_mode' => 'HTML4_Table'));
    } else {
        echo '<pre style="background-color:#eee; color:#000; padding:1em;">';
        var_dump($var);
        echo '</pre>';
    }
}

/**
 * Callback error handler for PEAR_PackageFileManager_Frontend and
 * its error stack.
 * @ignore
 */
function haltOnError($err)
{
    if ($err['level'] == 'error') {
        echo __FUNCTION__;
        varDump($err);
        return PEAR_ERRORSTACK_DIE;
    }
}

session_start();

// where to find package sources
$pkgDir = 'c:/php/pear/PEAR_PackageFileManager';

PEAR_ErrorStack::staticPushCallback('haltOnError');

$pfmfe =& PEAR_PackageFileManager_Frontend::singleton('Null', $pkgDir);
$pfmfe->setOption('baseinstalldir', 'PEAR');

echo '<h1>All maintainers</h1>';
$maint = $pfmfe->getMaintList();
varDump($maint);

echo '<h1>Leader only </h1>';
$maint = $pfmfe->getMaintList('lead');
varDump($maint);

echo '<h1>Developper only </h1>';
$maint = $pfmfe->getMaintList('developer');
varDump($maint);

echo '<h1>unknown users </h1>';
$maint = $pfmfe->getMaintList('users');
varDump($maint);

// cleansweep session data
$pfmfe->container(true);
session_destroy();
?>